### ISSUE TITLE
.on('end') runs before the last row of .on('data')

### ISSUE THREAD

**Issue by Nakarmi23** [C2FO/fast-csv#493]
The end event of the packages runs before the last row of .on('data').

**Comment by dany-fu** [comment]
@Nakarmi23 Are you using pause and resume because you're waiting for results from an async function? I stumbled on this because I thought I was having the same issue but it turns out that the value just wasn't being returned in time. Using `Promise.all()`resolved it

**Comment by Nakarmi23** [comment]
Yes i am using pause and resume to wait for an async value. On each loop of .on('data') i am performing some actions related to the database.

**Comment by doug-martin** [comment]
@Nakarmi23 I would assume @dany-fu is correct, when using async actions in your callback that _could_ happen because the stream has finished reading all of the data from the source but your code is still processing. Internally we did start with `pause` and `resume` but switched to [implementing our own `TransFormStreams`](https://nodejs.org/dist/latest-v14.x/docs/api/stream.html#stream_implementing_a_transform_stream) as they are a lot more predictable.

**Comment by ezfe** [comment]
I've also encountered this issue - I was doing database work in each `.on('data')`. I tried using `parser.pause()` but `.on('end')` still fired even though the parser was paused, causing my work to resolve before the final line was parsed.

**Comment by jadware** [comment]
this pattern should work:

`const stream = fs.createReadStream(filename)
	.pipe(csv.parse({ headers: true }))
	.on('data', row => promises.push(parseRow(row)))
	.on('end', async num_rows =>
	{
		await Promise.all(promises);
	});

async function parseRow(row)
{
	...
}
`

**Comment by grimen** [comment]
I have in middle of this issue now for the past 2 hours, tried to port older version stream API of this library to the new stream API. This is clearly broken in the internals, it behaves nothing like other stream APIs for similar scenarios (XML, JSON, etc) I have used past years. Kind of giving up for now.

**Comment by zwezheng** [comment]
We had the same issue. The .on('end') function runs before the last row of .on('data'). 
Promise.all wouldn't help in our case, since 
1. Data we parsed have dependencies on each other(say we need to dedupe). 
2. We can't throw everything to DB layer because DB would have a large amount of write and read locks. And programmers need to know explicitly what can possibly go wrong.
3. DB WCUs are just too expensive for this kind of batches. We'd like our client to wait extra time on it.

**Comment by arimgibson** [comment]
@jadware 's solution was exactly what I was going for, but I reached the same conclusion that with my current dataset of ~102k rows, that would use too much memory. A workaround could be to remove each individual promise from that array after it's been resolved, but that doesn't sound effective in my opinion as it could have a significant impact on performance.

**Comment by ondruska** [comment]
not sure if this is fully relevant but I resorted to:

```
return new Promise((resolve) => {
        const data = [];
        csv
        .parseFile(file)
        .on("data", (row) => data.push({
                attr1: row["attr1"]
            })
        )
        .on("end", () => {
            resolve(data);
        });
    });
```

**Comment by arimgibson** [comment]
@ondruska The problem that started this thread is `.on("end")` firing before the end of _async_ operations completed during the `.on("data")` function. In your example, you're doing a synchronous operation which runs just fine. I like your solution and see what you're getting at, think you're just solving a different (but still valid) problem :stuck_out_tongue_winking_eye:

**Comment by dwil618** [comment]
I was having this issue to. Sharing my workaround which seems to be working. I was trying to avoid the `Promise.all()` approach, as I didn't want to load all the data into memory at once.
I went with this approach instead - tracking the running promises, and checking if they are completed before resolving during the "end" callback.

``` typescript
await new Promise((resolve, reject) => {
    const promiseTracker: Record<string, boolean> = {}; // Object to track running promises

    parseStream(dataStream, { headers: true })
        .on("data", (data) => {
            promiseTracker[data.entryId] = true; // A unique identifier for the data entry. Alternatively could use uuid package
            CallApiPromise
                .then(handleResponse)
                .finally(() => {
                    delete promiseTracker[data.entryId];
                });
        })
        .on("error", reject)
        .on("end", () => {
            let interval: NodeJS.Timeout | undefined = undefined;
            const checkPromisesCompleted = () => {
                if (Object.keys(promiseTracker).length === 0) {
                    if (interval)
                       clearInterval(interval);
                    resolve("Done");
                    return;
                }
            }
            checkPromisesCompleted();

            setInterval(checkPromisesCompleted, 200); // Check every 200ms if all promises data entries are processed
            setTimeout(() => { // Timout of 5 seconds before reject
                if (interval)
                   clearInterval(interval);
                reject("Timeout");
            }, 5000);
        });
});
```

**Comment by arimgibson** [comment]
That's a really nice solution and snippet @dwil618, thanks for sharing!!! Going to hang onto that for the future

**Comment by leeoniya** [comment]
in case people want to try something different (and maintained): https://github.com/leeoniya/uDSV
